### PR TITLE
modified present_task method to return audios with random presentation

### LINKED ISSUE #126: Present Task method does not return audio randomly
Logic needs to be added to present task method to return audios randomly

### PR DESCRIPTION
**Pull request by najamF18**
Tasks' audios are shuffled randomly and presented to user to record his annotation

Closes #126 

Screenshots:
image

### CODE REVIEW

**Comment by turian** [comment]
`ExperimentTypeTaskPresentation.task_presentation` is text that should be presented at the top of the page.

For 2AFC it would be nicer to have the reference in the middle and the second and third on a row below it. No text like "First and second are the same".

**Comment by najamF18** [comment]
> `ExperimentTypeTaskPresentation.task_presentation` is text that should be presented at the top of the page.
> 
> For 2AFC it would be nicer to have the reference in the middle and the second and third on a row below it. No text like "First and second are the same".

I did not quite understand the first point. We need to display task presentation on the page? That will be seen by user and we don't want that do we? And can you please explain the second point a little? if we cannot show text like this what else should be shown

> `ExperimentTypeTaskPresentation.task_presentation` is text that should be presented at the top of the page.
> 
> For 2AFC it would be nicer to have the reference in the middle and the second and third on a row below it. No text like "First and second are the same".

should the annotations appear like this?
image

**Comment by najamF18** [comment]
> > `ExperimentTypeTaskPresentation.task_presentation` is text that should be presented at the top of the page.
> > For 2AFC it would be nicer to have the reference in the middle and the second and third on a row below it. No text like "First and second are the same".
> 
> > `ExperimentTypeTaskPresentation.task_presentation` is text that should be presented at the top of the page.
> > For 2AFC it would be nicer to have the reference in the middle and the second and third on a row below it. No text like "First and second are the same".
> 
> should the annotations appear like this? image

And about the arrangement, we need to show reference_url audio at top and then the last two below it. Right?

**Comment by najamF18** [comment]
> 

@turian please have a look at this:
image

**Comment by DeFidelity** [comment]
I think you need to update the image in the description

**Comment by najamF18** [comment]
> I think you need to update the image in the description

I have updated it.

**Comment by DeFidelity** [comment]
We might also need a little description of what got changed

File: polls/templates/polls/htmlform.html [hunk 1]
```diff
@@ -9,24 +9,38 @@ <h3>Task</h3>
...
+  <div>
+    {% for audio in audios %}
+
+    {% if forloop.counter == 1 %}
+    <div>
+
+      <audio controls src="{{ audio }}" class="mt-3">
+        Your browser does not support the
+        <code>audio</code> element.
+      </audio>
```
Comment (DeFidelity) [inline polls/templates/polls/htmlform.html:21]: I think the name of this file is too random too, can we use task-flow-form or something else that has to do with what the form does?

File: polls/templates/polls/htmlform.html [hunk 2]
```diff
@@ -9,24 +9,38 @@ <h3>Task</h3>
...
     </div>
-    <div class="form-check form-check-inline">
-      <input class="form-check-input" type="radio" id="type2" name="annotationOption" value="XYX">
-      <label class="form-check-label" for="type2">First and Third is same</label>
+
+    {% else %}
+    <audio controls src="{{ audio }}" class="mt-3">
+      Your browser does not support the
```
Comment (DeFidelity) [inline polls/templates/polls/htmlform.html:26]: I've seen this text in the if block and now in the else too, is there a condition that warrants that someone's browser wouldn't support the audio? and maybe we can suggest browsers that would work.

File: polls/templates/polls/htmlform.html [hunk 3]
```diff
@@ -9,24 +9,38 @@ <h3>Task</h3>
   <input type="hidden" value="{{ task_list }}" name="tasklist">
   <input type="hidden" value="{{ task_presentation }}" name="taskPresentation">
   <h4>ID: {{task.pk}}</h4>
-  <audio controls src="{{ reference_url }}" class="mt-3">
```
Comment (DeFidelity) [inline polls/templates/polls/htmlform.html:12]: At line 5 where the form tag is, if we're using htmx, the best way to declare the form is:

<form hx-post="{% url 'create_annotation' %}" hx-headers='{"X-CSRFToken": "{{ csrf_token }}"}' hx-boost="true">

File: polls/templates/polls/htmlform.html [hunk 4]
```diff
@@ -9,24 +9,38 @@ <h3>Task</h3>
...
+
+  <div class="mt-2 d-flex justify-content-around">
+    {% for annotation in task_annotations %}
+    <div class="form-check form-check-inline mx-10">
+      <input class="form-check-input" type="radio" id="type1" name="annotationOption"
+        value="{{ annotation.annotation }}">
     </div>
+    {% endfor %}
+
   </div>
```
Comment (DeFidelity) [inline polls/templates/polls/htmlform.html:45]: We can as well have hx-target="body" in the submit button tag, so the new page sent by the django views would be rendered on the body tag in base.

File: polls/templates/polls/htmlform.html (hunk 2 above)
Comment (najamF18) [reply polls/templates/polls/htmlform.html:26]: It is just for triangular view of audios.

File: polls/templates/polls/htmlform.html (hunk 1 above)
Comment (najamF18) [reply polls/templates/polls/htmlform.html:21]: sure we can change it

File: polls/templates/polls/htmlform.html (hunk 3 above)
Comment (najamF18) [reply polls/templates/polls/htmlform.html:12]: got it will change it

File: polls/templates/polls/htmlform.html (hunk 4 above)
Comment (najamF18) [reply polls/templates/polls/htmlform.html:45]: I think I may have a better approach for this. How about we just create a parent div and include the html returned by view in it. I can that will make the code more clean and lean

**Comment by najamF18** [comment]
> I think you need to update the image in the description

Have added it

**Review by DeFidelity — APPROVED** [review]

File: polls/utils.py [hunk 5]
```diff
@@ -14,11 +18,17 @@ def batch_selector():
...
+    task_presentations = ExperimentTypeTaskPresentation.objects.filter(
+        experiment_type=experiment_type
     )
-    return task.reference_url, task.transform_url, experiment_type.task_presentation
+    task_presentation_chosen = random.choice(task_presentations).task_presentation
+    unique_chars = set(task_presentation_chosen)
+    audio_mapping = {}
+    for index, key in enumerate(unique_chars):
+        audio_mapping[key] = audios[index]
+    return audio_mapping, task_presentation_chosen
```
Comment (turian) [inline polls/utils.py:31]: This is written in a really tricky way and TBH it's hard to prove this is correct. Especially given that there are ZERO comments and a later programmer might not understand it. I would prefer unit tests. And I would especially prefer that this were written in a really obvious easy-to-understand way that is easy to read and grok.

Comment (turian) [reply polls/utils.py:31]: Maybe you should add return types so this is more obvious

Why even return a dict audio_mapping? Why not just shuffle the URLs in the right order? So the functionality is totally contained in this method and encapsulated nicely and harder to break or mess up for another client?

Comment (najamF18) [reply polls/utils.py:31]: Ok let me try to refactor it in some easy way.

Comment (najamF18) [reply polls/utils.py]: > Maybe you should add return types so this is more obvious
> 
> Why even return a dict audio_mapping? Why not just shuffle the URLs in the right order? So the functionality is totally contained in this method and encapsulated nicely and harder to break or mess up for another client?

Actually we need to map those to know which audio refers which character like one audio will be bound to character "A" and the other one to "B". Now we can shuffle them according to task_presentation easily which could be anything like "AAB" or "BAB" etc

**Pull request merged**